graphql2/graphqlapp/alert.go

```diff
@@ -177,6 +179,85 @@ func (q *Query) mergeFavorites(ctx context.Context, svcs []string) ([]string, er
 	return svcs, nil
 }
 
+// splitRangeByDuration splits the timeframe given between since and until by the duration provided.
+// Each segment is then transformed into an AlertDataPoint based on the given alerts.
+// The given alerts are required to be sorted by their CreatedAt field.
+func splitRangeByDuration(since, until time.Time, dur timeutil.ISODuration, alerts []alert.Alert) (result []graphql2.AlertDataPoint) {
+	if since.After(until) {
+		return result
+	}
+
+	i := 0
+	// ignore alerts created before since
+	for i < len(alerts) && alerts[i].CreatedAt.Before(since) {
+		i++
+	}
+
+	ts, upperBound := since, since.AddDate(dur.Years, dur.Months, dur.Days).Add(dur.TimePart)
+	if upperBound.After(until) {
+		upperBound = until
+	}
+
+	for ts.Before(until) {
+		next := graphql2.AlertDataPoint{Timestamp: ts, AlertCount: 0}
+		for i < len(alerts) && alerts[i].CreatedAt.Before(upperBound) {
+			next.AlertCount++
+			i++
+		}
+		result = append(result, next)
+		ts, upperBound = upperBound, upperBound.AddDate(dur.Years, dur.Months, dur.Days).Add(dur.TimePart)
+		if upperBound.After(until) {
+			upperBound = until
+		}
+	}
```

Comment:
A lot of the logic makes sense, but this piece took me a little bit to figure out. So it seems you make a period, then fill in the alert count for that period, then make the next period, then fill in the alert count, etc etc.

Could this maybe be done more clearly by making all the periods first in one pass, then coming back and filling in all the data after? Especially since I'm sure we'll be adding to this in the future if we want to add more metrics.